### PR TITLE
Interface contracts

### PR DESCRIPTION
**Pull request by LefterisJP**
- Adding the possibility of omitting a function body by simply ending a
  function definition with a semicolon

- Such a function is marked as not fully implemented and any contract
  that contains such a function is considered a not fully implemented contract

### CODE REVIEW

File: libsolidity/AST.h [hunk 1]
```diff
@@ -224,7 +244,7 @@ class ContractDefinition: public Declaration, public Documented
 		m_definedFunctions(_definedFunctions),
 		m_functionModifiers(_functionModifiers),
 		m_events(_events)
-	{}
+		{}
```
Comment (chriseth) [inline libsolidity/AST.h]: I think this should not be indented.

File: libsolidity/AST.h [hunk 2]
```diff
@@ -378,25 +398,30 @@ class ParameterList: public ASTNode
...
+		Declaration(_location, _name, _visibility),
+		Documented(_documentation),
+		ImplementationOptional(_body != nullptr),
+		m_isConstructor(_isConstructor),
+		m_parameters(_parameters),
+		m_isDeclaredConst(_isDeclaredConst),
+		m_functionModifiers(_modifiers),
+		m_returnParameters(_returnParameters),
+		m_body(_body)
+		{}
```
Comment (chriseth) [inline libsolidity/AST.h]: Also here.

File: libsolidity/Parser.cpp [hunk 3]
```diff
@@ -116,6 +116,7 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
 {
 	ASTNodeFactory nodeFactory(*this);
 	ASTPointer<ASTString> docString;
+	bool contractFullyImplemented = true;
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: For the contract, this should be a type information:
If a contract does not overwrite all non-implemented functions from the base, it is considered "abstract".
These contracts cannot be compiled to a "final contract" and cannot be used in "new" expressions, which should be checked at compilation time.

File: libsolidity/AST.h [hunk 4]
```diff
@@ -378,24 +398,29 @@ class ParameterList: public ASTNode
...
+		ASTPointer<ASTString> const& _documentation,
+		ASTPointer<ParameterList> const& _parameters,
+		bool _isDeclaredConst,
+		std::vector<ASTPointer<ModifierInvocation>> const& _modifiers,
+		ASTPointer<ParameterList> const& _returnParameters,
+		ASTPointer<Block> const& _body
+	):
+		Declaration(_location, _name, _visibility),
+		Documented(_documentation),
+		ImplementationOptional(_body != nullptr),
```
Comment (chriseth) [inline libsolidity/AST.h]: `_body` and `nullptr` have different type. Better to use `!!_body`?

File: libsolidity/Parser.cpp [hunk 5]
```diff
@@ -141,7 +141,10 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
 		if (currentToken == Token::RBrace)
 			break;
 		else if (currentToken == Token::Function)
-			functions.push_back(parseFunctionDefinition(name.get()));
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: Why did you change that?

File: libsolidity/Parser.cpp [hunk 6]
```diff
@@ -247,8 +259,15 @@ ASTPointer<FunctionDefinition> Parser::parseFunctionDefinition(ASTString const*
...
-	nodeFactory.setEndPositionFromNode(block);
+	ASTPointer<Block> block = ASTPointer<Block>();
+	nodeFactory.markEndPosition();
+	if (m_scanner->getCurrentToken() != Token::Semicolon)
+	{
+		block = parseBlock();
+		nodeFactory.setEndPositionFromNode(block);
+	}
+	else
+		m_scanner->next(); // just consume the ';'
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: Could go without comment if you write `expectToken(Token::Semicolon)`.

File: libsolidity/Parser.cpp (hunk 5 above)
Comment (LefterisJP) [inline libsolidity/Parser.cpp]: incremental changes and reverts do that :P

File: libsolidity/Parser.cpp (hunk 6 above)
Comment (LefterisJP) [inline libsolidity/Parser.cpp]: yeah but that's an extra if check for something we just checked 3 lines above. It would be more readable .. but yeah, don't like to have useless checks.

**Comment by LefterisJP** [comment]
Implemented what we talked about

File: libsolidity/AST.cpp [hunk 7]
```diff
@@ -124,6 +130,42 @@ FunctionDefinition const* ContractDefinition::getFallbackFunction() const
 	return nullptr;
 }
 
+void ContractDefinition::checkAbstractFunctions()
+{
+	map<string, bool> functions;
+
+	// Search from base to derived
+	for (ContractDefinition const* contract: boost::adaptors::reverse(getLinearizedBaseContracts()))
+	{
```
Comment (chriseth) [inline libsolidity/AST.cpp]: No braces.

File: libsolidity/AST.cpp [hunk 8]
```diff
@@ -124,6 +130,42 @@ FunctionDefinition const* ContractDefinition::getFallbackFunction() const
...
+		for (ASTPointer<FunctionDefinition> const& function: contract->getDefinedFunctions())
+		{
+			string const& name = function->getName();
+			if (!function->isFullyImplemented() && functions.count(name) && functions[name])
+				BOOST_THROW_EXCEPTION(function->createTypeError("Redeclaring an already implemented function as abstract"));
+			// if (functions.count(name) && !functions[name] && function->isFullyImplemented())
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Can the comments be removed?

File: libsolidity/AST.cpp [hunk 9]
```diff
@@ -124,6 +130,42 @@ FunctionDefinition const* ContractDefinition::getFallbackFunction() const
...
+			// 	abs_functions.insert(make_pair(name, function.get()));
+		}
+	}
+	for (auto const& it: functions)
+		if (!it.second)
+		{
+			setFullyImplemented(false);
+			break;
+		}
+
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Remove blank line.

File: libsolidity/AST.cpp [hunk 10]
```diff
@@ -639,6 +681,8 @@ void NewExpression::checkTypeRequirements()
 	m_contract = dynamic_cast<ContractDefinition const*>(m_contractName->getReferencedDeclaration());
 	if (!m_contract)
 		BOOST_THROW_EXCEPTION(createTypeError("Identifier is not a contract."));
+	if (!m_contract->isFullyImplemented())
+		BOOST_THROW_EXCEPTION(m_contract->createTypeError("Trying to create an object of an abstract contract."));
```
Comment (chriseth) [inline libsolidity/AST.cpp]: `s/object/instance/`?

File: test/SolidityNameAndTypeResolution.cpp [hunk 11]
```diff
@@ -346,6 +346,63 @@ BOOST_AUTO_TEST_CASE(comparison_bitop_precedence)
...
+}
+
+BOOST_AUTO_TEST_CASE(create_abstract_contract)
+{
+	ASTPointer<SourceUnit> sourceUnit;
+	char const* text = R"(
+		contract base { function foo(); }
+		contract derived { 
+			base b;
+            function foo() { b = new base();}
```
Comment (chriseth) [inline test/SolidityNameAndTypeResolution.cpp]: Indentation

File: libsolidity/AST.cpp (hunk 8 above)
Comment (LefterisJP) [inline libsolidity/AST.cpp]: I think that would be a good idea ... no idea how they ended up staying there ....
